### ISSUE TITLE
Using multi dimensional arrays with different orders to create another array, causes a semantics error

### ISSUE THREAD

**Issue by ohachim** [pyccel/pyccel#1170]
**To Reproduce**
```python
if __name__ == "__main__":
    import numpy as np
    a = np.array([[1], [2], [4]], order="F")
    b = np.array([[4], [5], [6]], order="C")
    c = np.array([a, b], order="C")
```
```a``` and ```b``` must be multi dimensional and of different orders to reproduce this bug.

Generates a semantics error
```ERROR at annotation (semantic) stage
pyccel:
 |fatal [semantic]: testbug.py [5,17]| Inhomogeneous lists are not supported by Pyccel. Please use a tuple ((a, b))
```

**Expected behavior**
Should pass the semantics check.

**Additional context**
As long as it passes the semantics check, Pyccel will be able to copy Python's behaviour.

**Comment by azouiten** [comment]
I don't know if this is a bug, in the Python tuple class init checks for the order to decide if its the same type
`is_homogeneous = arg0.dtype is not NativeGeneric() and \
                         all(a.dtype is not NativeGeneric() and \
                             arg0.dtype == a.dtype and \
                             arg0.rank  == a.rank  and \
                             arg0.order == a.order for a in args[1:])`

**Comment by ohachim** [comment]
@azouiten I see, but I figured that since it works in python, it should work here, and I thought that it wouldn't be hard to implement, actually, in the implementation am working on, it will work.
@bauom @EmilyBourne

**Comment by EmilyBourne** [comment]
I agree with @azouiten partially. Currently pyccel does not support lists and when it does it will almost certainly not support inhomogeneous lists. Currently the example raises an error, explains why and explains how to get almost identical code that works. You can't do much better than that if you don't want to support something. Pyccel will never do everything python can otherwise it would be as slow as python.

The reason we support inhomogeneous tuples but not lists is due to the immutable nature of tuples. You can translate
```
a = (1, True, 7.9)
```
if you know that it will always contain those values but one of the things that makes lists special is that you can append and modify elements.
Eg
```
a = [1, True, 7.9]
...
a[i] = False
```
Obviously the code above cannot be translated as the type of an object may be being changed during execution.
So we could support lists if
- their elements never change (this is equivalent to asking people to use tuples instead)
- elements are never accessed via variables (this seems like quite an unreasonable restriction for lists. It could be applied only to inhomogeneous lists but that may be hard to indicate clearly

Supporting inhomogeneous lists would also mean supporting statements like this:
```
a= []
if b:
    a.append(False)
else:
    a.append(3.14)
```

**Comment by EmilyBourne** [comment]
It is easy to handle special cases, but it is important to consider the consequences of deciding to support new constructs

**Comment by ohachim** [comment]
But the order of an array is basically just how it is placed in the memory, the data type is the same, handling diffrent orders should't have special cases (as far as I know)

**Comment by EmilyBourne** [comment]
> But the order of an array is basically just how it is placed in the memory, the data type is the same, handling diffrent orders should't have special cases (as far as I know)

That makes sense. But then you allow:

```
a = []

for i in range(10):
     b = calculate_b()
     if b:
          a.append(np.ones((3,4), order='C'))
     else
          a.append(np.ones((3,4), order='F'))
```
How do you access the elements of a[7] after the loop?

**Comment by ohachim** [comment]
Hmmm,  I think I am missing something, wouldn't we just use the strides of a[7]? if a[7] is of type ```ndarray```?

**Comment by EmilyBourne** [comment]
In Fortran the strides are not multiplied by the other dimensions. The ordering just tells us whether the indexing should be treated as column major or row major. (Someone needs to write a section in the developer docs about this)

Have a look at the Fortran translation of:
```
def f():
    a = np.ones((4,6), order='F')
    for ai in a:
        print(ai)

def g():
    a = np.ones((4,6), order='C')
    for ai in a:
        print(ai)
```
And see if you can see how it works

**Comment by ohachim** [comment]
I understand, thanks!

**Comment by azouiten** [comment]
I thought that inhomogeneous tuples are not supported given that pyccel code breaks when trying to give it a heterogeneous tuple,  it gives this error:
`
TypeError: can only concatenate tuple (not "NoneType") to tuple
`
@EmilyBourne

**Comment by EmilyBourne** [comment]
> I thought that inhomogeneous tuples are not supported given that pyccel code breaks when trying to give it a heterogeneous tuple, it gives this error: `TypeError: can only concatenate tuple (not "NoneType") to tuple` @EmilyBourne

What code were you trying to translate?

**Comment by EmilyBourne** [comment]
There are lots of tests for heterogeneous tuples [here](https://github.com/pyccel/pyccel/blob/master/tests/epyccel/modules/tuples.py)

**Comment by azouiten** [comment]
```python
if __name__ == "__main__":
    a = (1, [1, 2])
```

**Comment by EmilyBourne** [comment]
> I understand, thanks!

If you have understood how it works then it would be great if you could add the docs about the "order". That would also allow me to reread it and make sure you have understood correctly.

**Comment by EmilyBourne** [comment]
> I tried this
> 
> ```python
> if __name__ == "__main__":
>     a = (1, [1, 2])
> ```

Hmm, this looks like a bug. I suspect that the homogeneous test doesn't check the rank and therefore assumes this is homogeneous. But it can't be stored in a homogeneous object as it doesn't have a fixed rank

**Comment by azouiten** [comment]
make sense, I ll open an issue for it.

**Comment by EmilyBourne** [comment]
Currently tuples are split into 2 types:

## Homogeneous

Internally this works like an array. The advantage of this is that it can be indexed with a variable and it can be iterated over.
The disadvantage is that the homogeneity needs to be checked correctly (I don't think it is currently)

## Inhomogeneous tuples

Internally this works like multiple separate variables.
I.e:
```python
a = (1, [1, 2])
```
should be equivalent to
```python
a_0 = 1
a_1 = [1, 2]
```
The advantage is that we can handle the hidden variables as we would normal variables.
We can't index with a variable, but this wouldn't be possible anyway as we can't know the type in this case.

## Open Question

While writing:
```python
a = (1, 2, 4 ,6, 10, 134)
```
as a homogeneous tuple makes sense, it's not clear if this does:
```python
a = ((1,3), (2,-2), (4 ,6), (10, 134))
```
But if it doesn't and we should reduce our support back to just tuples treated as inhomogeneous objects then the question is how to explain this to users, or whether we should completely drop support for homogeneous objects

**Comment by EmilyBourne** [comment]
> I don't understand why this example would be a problem
> 
> ```python
> a = ((1,3), (2,-2), (4 ,6), (10, 134))
> ```
> 
> could you please elaborate.

It is not a problem, and I think it works currently. But it only works because we know that the shape of all these elements is the same. I'm not sure that it is very transparent for a user why:
```python
a = ((1,3), (2,-2), (4 ,6), (10, 134))
```
should work but 
```python
a = ((1,3), (2,-2), (4 ,6), (10, 134, -4))
```
shouldn't

**Comment by azouiten** [comment]
but even the second example would work if there is support for inhomogeneous tuples, I think that like you said the current problem is that we currently don't classify the objects correctly (homogeneous/heterogeneous). unless the correct classification would bring up ambiguities or a lot of specific cases to handle.

**Comment by EmilyBourne** [comment]
> but even the second example would work if there is support for inhomogeneous tuples, I think that like you said the problem is that we currently don't classify the objects correctly (homogeneous/heterogeneous). unless the correct classification would bring up ambiguities or a lot of specific cases to handle.

Yes you are right. But the term "homogeneous" usually refers to types, i.e. things that would need different type descriptors in fortran.
This is not the case for `a` and `b` here, so it may be hard for a user to understand why it is considered to be heterogeneous.
In C it could be implemented homogeneously by doing:
```
ndarray a;
ndarray b;
ndarray* c[2];
c[0] = &a;
c[1] = &b;
```

**Comment by azouiten** [comment]
I think I understand now, could we add language specific attributes (attributes that are intended only for a specific language printing ) to these objects that would specify different writing functions for both C and Fortran. I don't know if that would work.

**Comment by EmilyBourne** [comment]
> I think I understand now, could we add language specific attributes (attributes that are intended only for a specific language printing ) to these objects that would specify different writing functions for both C and Fortran. I don't know if that would work.

We talked about adding a language-specific semantic stage at some point. At the moment we try to keep the implementations as similar as possible and add any language-specific minor changes in the code printing stage.

In any case, just because you can do:
```python
ndarray a;
ndarray b;
ndarray* c[2];
c[0] = &a;
c[1] = &b;
```
it doesn't mean it will be efficient or a good idea. Sometimes pyccel restrictions are just a way of encouraging users to follow best practices which can also accelerate their code
